subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java

```diff
@@ -85,15 +81,16 @@ private void walkNestedBean(T node, TypeMetadata typeMetadata, String qualifiedN
 
     private void walkChildren(T node, TypeMetadata typeMetadata, @Nullable String parentQualifiedName, V visitor, Map<T, String> nestedNodesOnPath) {
         typeMetadata.getPropertiesMetadata().forEach(propertyMetadata -> {
-            String childQualifiedName = getQualifiedName(parentQualifiedName, propertyMetadata.getPropertyName());
             if (propertyMetadata.getPropertyType() == nestedAnnotation) {
-                walkNestedChild(node, childQualifiedName, propertyMetadata, visitor, child -> walkNested(child, childQualifiedName, propertyMetadata, visitor, nestedNodesOnPath, false));
+                walkNestedChild(node, getQualifiedName(parentQualifiedName, propertyMetadata.getPropertyName()), propertyMetadata, visitor, child -> walkNested(child, getQualifiedName(parentQualifiedName, propertyMetadata.getPropertyName()), propertyMetadata, visitor, nestedNodesOnPath, false));
             } else {
-                visitor.visitLeaf(childQualifiedName, propertyMetadata, () -> getChild(node, propertyMetadata));
+                walkLeaf(node, parentQualifiedName, visitor, propertyMetadata);
             }
         });
     }
 
+    abstract protected void walkLeaf(T node, @Nullable String parentQualifiedName, V visitor, PropertyMetadata propertyMetadata);
```

Comment:
```suggestion
    abstract protected void walkLeaf(T parent, @Nullable String parentQualifiedName, V visitor, PropertyMetadata propertyMetadata);
```

🧐 Maybe rename it to parent.

Reply:
As discussed below, this is reverted in https://github.com/gradle/gradle/pull/22788.